### DISCUSSION TITLE
Poor performance for generating UMA tickets for a scope permission

### DISCUSSION THREAD

**Discussion by krystiansola** [keycloak/keycloak#30668] [Q&A]
The response time for generating an access token for a UMA ticket takes 2 minutes or sometimes the response time does not succeed at all.

What am I doing:
We have a resource of type: `workspace`. In keycloak we have over 4000 resources of that type. For every resource we have multiple permissions. The number is between 1 to 20 permissions for each resource. Every resource has a VIEW Scope

Now I want check to which workspace I have the VIEW scope.

I do the following request:

curl --request POST \
  --url https://[url-to-keycloak]/auth/realms/cs/protocol/openid-connect/token \
  --header 'Authorization: Basic <credentials-for-client>’ \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data audience=[client-id] \
  --data grant_type=urn:ietf:params:oauth:grant-type:uma-ticket \
  --data subject_token=[my-access-token]
  —permission=#VIEW

This request works from 2-10 min and it is possible that the request does not succeed at all. 

Why this request takes so long? The number 4000 seems not be a big number of resources. 

The permission which I am creating is a scope base permission with the resource id and Scope VIEW. It can be that we have over 40’000 permissions

Am I doing something against the design of keycloak?

What I want achieve that in a every workspace we have an:
- ADMIN (has 5 different scopes)
- EDITOR(has 4 different scopes) 
- VIEWER (has 1 scope)

So for every ADMIN I create a scope permission where we set the resource id (which equals workspaceId), scopes for admin and user policy with a user id of the chosen admin. The same wa are doing for EDITOR and VIEWER.

Example of such permission:
<img width="1018" alt="Zrzut ekranu 2024-06-22 o 09 15 17" src="https://github.com/keycloak/keycloak/assets/161840690/7fe0fa35-c1df-4436-aa73-ff240ed0ae9b">

**Comment by jonkoops** [comment]
From the screenshot you posted here I can assume this is an incredibly outdated version of Keycloak. I would highly recommend you first update to the latest version of Keycloak, as it will increase your chances of getting support, as well as likely fixing this issue if it was a known one. Do be aware that running an outdated version of Keycloak means you are vulnerable to several critical security vulnerabilities.

**Comment by krystiansola** [comment]
@jonkoops Thank you for the answer.

We have keycloak 17. Now we will upgrade to keycloak 18. 

Now we have the process of upgrading keycloak by 1 version. I wondering if this is necessary. Do you have any best practice? Would it make sense to upgrade to keycloak 25?

Regards the resources and number of permission: is in your opinion reasonable the design to create so many permissions for one resources? Here I am not sure if I made the right design decisions. Any opinions on that?

**Reply by jonkoops** [reply]
> Would it make sense to upgrade to keycloak 25?

Yes, officially we only support the latest stable version of Keycloak, so my recommendation would be to always upgrade to the latest if possible.

> Regards the resources and number of permission: is in your opinion reasonable the design to create so many permissions for one resources? Here I am not sure if I made the right design decisions. Any opinions on that?

I am not familiar with the scalability of this, so I can't help you there unfortunately, but considering the performance currently I'd advise to reduce this amount of it. @keycloak/core might be able to answer this.

**Reply by krystiansola** [reply]
But would you make a jump from 17 directly to 25 or upgrade version by version

**Reply by jonkoops** [reply]
My recommendation would be to go major by major rather than all at once. There can be migration steps between versions that might be to big to handle otherwise. I'd also recommend using the latest version of each major to avoid any unwanted issues (e.g. `24.0.5` and not `24.0.0`).

**Reply by krystiansola** [reply]
Thank you. We will upgrade and than again check the performance.

**Reply by jonkoops** [reply]
Great, if the performance is an issue do let us know :+1:
